Documentation fails to build on docs.rs

See the build log for the newly released `0.2.7`: https://docs.rs/crate/adblock/0.2.7/builds/237004

It contains the error message:
```
[INFO] [stderr] error: proc-macro derive panicked
[INFO] [stderr]  --> /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/psl-0.4.1/src/list.rs:6:10
[INFO] [stderr]   |
[INFO] [stderr] 6 | #[derive(Psl, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
[INFO] [stderr]   |          ^^^
[INFO] [stderr]   |
[INFO] [stderr]   = help: message: failed to download the list: failed to lookup address information: Temporary failure in name resolution
[INFO] [stderr] 
[INFO] [stderr] error[E0599]: no method named `domain` found for struct `list::List` in the current scope
[INFO] [stderr]   --> /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/psl-0.4.1/src/list.rs:28:20
[INFO] [stderr]    |
[INFO] [stderr] 7  | pub struct List;
[INFO] [stderr]    | ---------------- method `domain` not found for this
[INFO] [stderr] ...
[INFO] [stderr] 28 |         match List.domain(input) {
[INFO] [stderr]    |                    ^^^^^^ method not found in `list::List`
[INFO] [stderr]    |
[INFO] [stderr]    = help: items from traits can only be used if the trait is implemented and in scope
[INFO] [stderr] note: `Psl` defines an item `domain`, perhaps you need to implement it
[INFO] [stderr]   --> /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/psl-0.4.1/src/lib.rs:48:1
[INFO] [stderr]    |
[INFO] [stderr] 48 | pub trait Psl {
[INFO] [stderr]    | ^^^^^^^^^^^^^
[INFO] [stderr] 
[INFO] [stderr] error[E0599]: no method named `suffix` found for struct `list::List` in the current scope
[INFO] [stderr]   --> /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/psl-0.4.1/src/list.rs:51:20
[INFO] [stderr]    |
[INFO] [stderr] 7  | pub struct List;
[INFO] [stderr]    | ---------------- method `suffix` not found for this
[INFO] [stderr] ...
[INFO] [stderr] 51 |         match List.suffix(input) {
[INFO] [stderr]    |                    ^^^^^^ method not found in `list::List`
[INFO] [stderr]    |
[INFO] [stderr]    = help: items from traits can only be used if the trait is implemented and in scope
[INFO] [stderr] note: `Psl` defines an item `suffix`, perhaps you need to implement it
[INFO] [stderr]   --> /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/psl-0.4.1/src/lib.rs:48:1
[INFO] [stderr]    |
[INFO] [stderr] 48 | pub trait Psl {
[INFO] [stderr]    | ^^^^^^^^^^^^^
[INFO] [stderr] 
[INFO] [stderr] error: aborting due to 3 previous errors
[INFO] [stderr] 
[INFO] [stderr] For more information about this error, try `rustc --explain E0599`.
[INFO] [stderr] error: could not compile `psl`.
[INFO] [stderr] 
[INFO] [stderr] To learn more, run the command again with --verbose.
[INFO] [stderr] warning: build failed, waiting for other jobs to finish...
[INFO] [stderr] error: build failed
[INFO] running `"docker" "inspect" "1a583d3b205a1d35ca294c083e029718f790f76ef98f367be6181f879fdfafe8"`
[INFO] running `"docker" "rm" "-f" "1a583d3b205a1d35ca294c083e029718f790f76ef98f367be6181f879fdfafe8"`
[INFO] [stdout] 1a583d3b205a1d35ca294c083e029718f790f76ef98f367be6181f879fdfafe8
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -19,15 +19,15 @@ exclude = [
 
 [package.metadata.docs.rs]
 no-default-features = true
-features = ["default-not-docs-rs"]
+features = ["default-docs-rs"]
 
 [dependencies]
-addr = "0.2"
+addr = { version = "0.2", optional = true }
 url = "2.1"
 percent-encoding = "2.1"
 matches = "0.1"
 lazy_static = "1.4"
-once_cell = "1.2"
+once_cell = "1.4"
 regex = "1"
 bitflags = "1.2"
 itertools = "0.8"
@@ -79,13 +79,14 @@ name = "bench_cosmetic_matching"
 harness = false
 
 [features]
-# If disabling default features, be sure to explicitly re-enable the
-# "docs-rs-incompatible" feature, or else the library will not work correctly.
-default = ["default-not-docs-rs", "docs-rs-incompatible"]
-default-not-docs-rs = ["full-regex-handling", "object-pooling"]
+# If disabling default features, consider explicitly re-enabling the
+# "embedded-domain-resolver" feature.
+default = ["default-docs-rs", "docs-rs-incompatible"]
+default-docs-rs = ["full-regex-handling", "object-pooling"]
 full-domain-matching = []
 metrics = []
 full-regex-handling = []
 object-pooling = ["lifeguard"]
 css-validation = ["cssparser", "selectors"]
-docs-rs-incompatible = [] # psl prevents docs.rs builds, see https://github.com/rust-lang/docs.rs/issues/904
+embedded-domain-resolver = ["addr"] # Requires setting an external domain resolver if disabled.
+docs-rs-incompatible = ["embedded-domain-resolver"] # addr prevents docs.rs builds, see https://github.com/rust-lang/docs.rs/issues/904

--- a/native/Cargo.toml
+++ b/native/Cargo.toml
@@ -16,5 +16,5 @@ neon-build = "0.3.3"
 [dependencies]
 neon-serde = "0.3"
 serde =  { version = "1.0", features = ["derive", "rc"] }
-adblock = { path = "../", default-features = false, features = ["css-validation", "docs-rs-incompatible"] }
+adblock = { path = "../", default-features = false, features = ["css-validation", "embedded-domain-resolver"] }
 neon = "0.3.3"

--- a/src/request.rs
+++ b/src/request.rs
@@ -1,4 +1,4 @@
-use crate::url_parser::{get_host_domain, UrlParser};
+use crate::url_parser;
 use crate::utils;
 
 use idna;
@@ -217,8 +217,8 @@ impl<'a> Request {
         source_url: &str,
         request_type: &str,
     ) -> Result<Request, RequestError> {
-        if let Some(parsed_url) = Request::parse_url(&url) {
-            if let Some(parsed_source) = Request::parse_url(&source_url) {
+        if let Some(parsed_url) = url_parser::parse_url(&url) {
+            if let Some(parsed_source) = url_parser::parse_url(&source_url) {
                 let source_domain = parsed_source.domain();
 
                 let third_party = if source_domain.is_empty() {
@@ -263,14 +263,14 @@ impl<'a> Request {
     ) -> Request {
         let url_norm = url.to_ascii_lowercase();
 
-        let (source_domain_start, source_domain_end) = get_host_domain(&source_hostname);
+        let (source_domain_start, source_domain_end) = url_parser::get_host_domain(&source_hostname);
         let source_domain = &source_hostname[source_domain_start..source_domain_end];
 
         let splitter = url_norm.find(':').unwrap_or(0);
         let schema: &str = &url[..splitter];
 
         let third_party = if third_party_request.is_none() {
-            let (domain_start, domain_end) = get_host_domain(&hostname);
+            let (domain_start, domain_end) = url_parser::get_host_domain(&hostname);
             let domain = &hostname[domain_start..domain_end];
             if source_domain.is_empty() {
                 None

--- a/src/url_parser/mod.rs
+++ b/src/url_parser/mod.rs
@@ -1,7 +1,72 @@
 mod parser;
-use crate::request::Request;
-use addr::DomainName;
 // mod parser_regex;
+
+#[cfg(not(feature = "embedded-domain-resolver"))]
+static DOMAIN_RESOLVER: once_cell::sync::OnceCell<Box<dyn ResolvesDomain>> = once_cell::sync::OnceCell::new();
+
+/// Sets the library's domain resolver implementation.
+///
+/// If the library is used without this having been set, panics may occur!
+///
+/// Will return the resolver if it has already been previously set.
+#[cfg(not(feature = "embedded-domain-resolver"))]
+pub fn set_domain_resolver(resolver: Box<dyn ResolvesDomain>) -> Result<(), Box<dyn ResolvesDomain>> {
+    DOMAIN_RESOLVER.set(resolver)
+}
+
+#[cfg(feature = "embedded-domain-resolver")]
+struct DefaultResolver;
+
+#[cfg(feature = "embedded-domain-resolver")]
+impl ResolvesDomain for DefaultResolver {
+    fn get_host_domain(&self, host: &str) -> (usize, usize) {
+        if host.is_empty() {
+            (0, 0)
+        } else {
+            match host.parse::<addr::DomainName>() {
+                Err(_e) => (0, host.len()),
+                Ok(domain) => {
+                    let root = domain.root();
+                    let domain_str = root.to_str();
+                    let domain_len = domain_str.len();
+                    let host_len = host.len();
+                    (host_len - domain_len, host_len)
+                }
+            }
+        }
+    }
+}
+
+pub trait ResolvesDomain: Send + Sync {
+    /// Return the start and end indices of the domain (eTLD+1) of the given hostname.
+    ///
+    /// If there isn't a valid domain, `(0, host.len())` should be returned.
+    ///
+    /// ```
+    /// # use adblock::url_parser::ResolvesDomain;
+    /// # /// I'd use DefaultResolver here, but I can't use private structs in doctests.
+    /// # /// Enjoy this mock implementation instead :(
+    /// # struct Resolver;
+    /// # impl ResolvesDomain for Resolver {
+    /// #     fn get_host_domain(&self, host: &str) -> (usize, usize) {
+    /// #         match host {
+    /// #             "api.m.example.com" => (6, 17),
+    /// #             "a.b.co.uk" => (2, 9),
+    /// #             _ => unreachable!()
+    /// #         }
+    /// #     }
+    /// # }
+    /// # let resolver = Resolver;
+    /// let host = "api.m.example.com";
+    /// let (start, end) = resolver.get_host_domain(host);
+    /// assert_eq!(&host[start..end], "example.com");
+    ///
+    /// let host = "a.b.co.uk";
+    /// let (start, end) = resolver.get_host_domain(host);
+    /// assert_eq!(&host[start..end], "b.co.uk");
+    /// ```
+    fn get_host_domain(&self, host: &str) -> (usize, usize);
+}
 
 pub struct RequestUrl {
     pub url: String,
@@ -22,45 +87,50 @@ impl RequestUrl {
     }
 }
 
-pub trait UrlParser {
-    /// Return the string representation of the host (domain or IP address) for this URL, if any together with the URL.
-    ///
-    /// As part of hostname parsing, punycode decoding is used to convert URLs with UTF characters to plain ASCII ones.
-    /// Serialisation then contains this decoded URL that is used for further matching.
-    fn parse_url(url: &str) -> Option<RequestUrl>;
+/// Return the start and end indices of the domain of the given hostname.
+pub(crate) fn get_host_domain(host: &str) -> (usize, usize) {
+    #[cfg(not(feature = "embedded-domain-resolver"))]
+    let domain_resolver = DOMAIN_RESOLVER.get().expect("An external domain resolver must be set when the `embedded-domain-resolver` feature is disabled.");
+    #[cfg(feature = "embedded-domain-resolver")]
+    let domain_resolver = DefaultResolver;
+
+    domain_resolver.get_host_domain(host)
 }
 
-impl UrlParser for Request {
-    fn parse_url(url: &str) -> Option<RequestUrl> {
-        let parsed = parser::Hostname::parse(&url).ok();
-        parsed.and_then(|h| {
-            match h.host_str() {
-                Some(_host) => Some(RequestUrl {
-                    url: h.url_str().to_owned(),
-                    schema_end: h.scheme_end,
-                    hostname_pos: (h.host_start, h.host_end),
-                    domain: get_host_domain(&h.url_str()[h.host_start..h.host_end])
-                }),
-                _ => None
-            }
-        })
-
-    }
-}
-
-pub fn get_host_domain(host: &str) -> (usize, usize) {
-    if host.is_empty() {
-        (0, 0)
-    } else {
-        match host.parse::<DomainName>() {
-            Err(_e) => (0, host.len()),
-            Ok(domain) => {
-                let root = domain.root();
-                let domain_str = root.to_str();
-                let domain_len = domain_str.len();
-                let host_len = host.len();
-                (host_len - domain_len, host_len)
-            }
+/// Return the string representation of the host (domain or IP address) for this URL, if any together with the URL.
+///
+/// As part of hostname parsing, punycode decoding is used to convert URLs with UTF characters to plain ASCII ones.
+/// Serialisation then contains this decoded URL that is used for further matching.
+pub fn parse_url(url: &str) -> Option<RequestUrl> {
+    let parsed = parser::Hostname::parse(&url).ok();
+    parsed.and_then(|h| {
+        match h.host_str() {
+            Some(_host) => Some(RequestUrl {
+                url: h.url_str().to_owned(),
+                schema_end: h.scheme_end,
+                hostname_pos: (h.host_start, h.host_end),
+                domain: get_host_domain(&h.url_str()[h.host_start..h.host_end])
+            }),
+            _ => None
         }
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    #[cfg(feature = "embedded-domain-resolver")]
+    fn test_get_host_domain() {
+        fn domain(host: &str) -> &str {
+            let resolver = DefaultResolver;
+            let (a, b) = resolver.get_host_domain(host);
+            &host[a..b]
+        }
+        assert_eq!(domain("www.google.com"), "google.com");
+        assert_eq!(domain("google.com."), "google.com.");
+        assert_eq!(domain("a.b.co.uk"), "b.co.uk");
+        assert_eq!(domain("foo.bar"), "foo.bar");
     }
 }

--- a/tests/ublock-coverage.rs
+++ b/tests/ublock-coverage.rs
@@ -2,8 +2,6 @@ extern crate adblock;
 
 use adblock::engine::Engine;
 use adblock::lists::FilterFormat;
-use adblock::request::Request;
-use adblock::url_parser::UrlParser;
 use adblock::utils::rules_from_lists;
 
 use serde::Deserialize;
@@ -220,8 +218,8 @@ fn check_matching_hostnames() {
     let engine = get_blocker_engine();
 
     for req in requests {
-        let url_host = Request::parse_url(&req.url).unwrap();
-        let source_host = Request::parse_url(&req.sourceUrl).unwrap();
+        let url_host = adblock::url_parser::parse_url(&req.url).unwrap();
+        let source_host = adblock::url_parser::parse_url(&req.sourceUrl).unwrap();
         let domain = url_host.domain();
         let source_domain = source_host.domain();
         let third_party = if source_domain.is_empty() {

